### PR TITLE
    Rework tool/doc for updating reference docs

### PR DESCRIPTION
**Pull request by tengqm**
This PR reworks/reimplements the reference docs generator in Python and fixes some outdated docs and data:
    
- Do docs import using Python because the GoLang version of tool has following drawbacks:
      * not convenient for handling YAML config files
      * it has to be compiled to binaries to run on different platforms
      * for every tiny changes you need to compile a new version and check in
- The reference docs we use in website are actually not coming directly from `kubernetes/kubernetes`. Most of them come from the `reference-docs` project. The configuration files are thus changed to avoid confusion.
- We have changed the location of generated docs so the default configuration files and the docs are updated.

Closes: #9003

### CODE REVIEW

**Comment by Bradamant3** [comment]
/assign @jimangel 

Sorry, should have put you on this one earlier!

File: content/en/docs/contribute/start.md [hunk 1]
```diff
@@ -28,7 +28,7 @@ The Kubernetes documentation is written in Markdown and processed and deployed
 using Hugo. The source is in Github at
 [https://github.com/kubernetes/website](https://github.com/kubernetes/website).
 Most of the documentation source is stored in `/content/en/docs/`. Some of the
-reference documentation is automatically generated from scripts, mostly in the
+reference documentation are automatically generated from scripts in the
```
Comment (lucperkins) [inline content/en/docs/contribute/start.md]: This should be "is"

File: content/en/docs/contribute/generate-ref-docs/kubernetes-components.md [hunk 2]
```diff
@@ -118,134 +118,79 @@ need to work with someone who can set the label and milestone for you.
...
+- name: reference-docs
+  remote: https://github.com/kubernetes-incubator/reference-docs.git
+  # This and the generate-command below needs a change when reference-docs has
+  # branches properly defined
+  branch: master  
+  generate-command: |
+    cd $GOPATH
+    git clone https://github.com/kubernetes/kubernetes.git src/k8s.io/kubernetes
+    cd src/k8s.io/kubernetes
+    git checkout release-1.11
```
Comment (jimangel) [inline content/en/docs/contribute/generate-ref-docs/kubernetes-components.md]: Do we want to make release-1.11 a variable?

**Comment by jimangel** [comment]
I followed this PR / branch for generating docs using the python script. I don't regularly use python so I hit some minor snags. 2 Nits:

- Error when I `go get` website
```
can't load package: package github.com/kubernetes/website: no Go files in $GOPATH/src/github.com/kubernetes/website
```
This makes sense, but maybe add a note that this can be ignored since k/website is not a true go project.

- Error running python script
```
$ ./update-imported-docs reference.yml
Traceback (most recent call last):
  File "./update-imported-docs", line 10, in <module>
    import yaml
ImportError: No module named yaml

# ADD TO PREREQS / SOLUTION:
sudo apt install python-pip
pip install pyyaml
```
I also found it strange that it didn't generate all of the `command-line-tools-reference` docs, but I was thinking that it might have been since I was using your branch for testing and they might have already been the latest.

Below is my output of what was modified:
```
git status
On branch test
Your branch is up to date with 'origin/update-doc-import'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   content/en/docs/reference/command-line-tools-reference/cloud-controller-manager.md
	modified:   content/en/docs/reference/command-line-tools-reference/federation-apiserver.md
	modified:   content/en/docs/reference/command-line-tools-reference/federation-controller-manager.md
	modified:   content/en/docs/reference/command-line-tools-reference/kube-controller-manager.md
	modified:   content/en/docs/reference/kubectl/kubectl.md
	modified:   content/en/docs/reference/setup-tools/kubeadm/generated/kubeadm_alpha_phase_addon_all.md
	modified:   content/en/docs/reference/setup-tools/kubeadm/generated/kubeadm_alpha_phase_addon_coredns.md
	modified:   content/en/docs/reference/setup-tools/kubeadm/generated/kubeadm_alpha_phase_addon_kube-proxy.md
	modified:   content/en/docs/reference/setup-tools/kubeadm/generated/kubeadm_alpha_phase_controlplane_all.md
	modified:   content/en/docs/reference/setup-tools/kubeadm/generated/kubeadm_alpha_phase_controlplane_apiserver.md
	modified:   content/en/docs/reference/setup-tools/kubeadm/generated/kubeadm_alpha_phase_controlplane_controller-manager.md
	modified:   content/en/docs/reference/setup-tools/kubeadm/generated/kubeadm_alpha_phase_controlplane_scheduler.md
	modified:   content/en/docs/reference/setup-tools/kubeadm/generated/kubeadm_config_images_list.md
	modified:   content/en/docs/reference/setup-tools/kubeadm/generated/kubeadm_config_images_pull.md
	modified:   content/en/docs/reference/setup-tools/kubeadm/generated/kubeadm_config_upload_from-flags.md
	modified:   content/en/docs/reference/setup-tools/kubeadm/generated/kubeadm_init.md
	modified:   content/en/docs/reference/setup-tools/kubeadm/generated/kubeadm_reset.md
	modified:   content/en/docs/reference/setup-tools/kubeadm/generated/kubeadm_token.md
	modified:   content/en/docs/reference/setup-tools/kubeadm/generated/kubeadm_upgrade.md
	modified:   content/en/docs/reference/setup-tools/kubeadm/generated/kubeadm_upgrade_apply.md
	modified:   content/en/docs/reference/setup-tools/kubeadm/generated/kubeadm_upgrade_diff.md
	modified:   content/en/docs/reference/setup-tools/kubeadm/generated/kubeadm_upgrade_node.md
	modified:   content/en/docs/reference/setup-tools/kubeadm/generated/kubeadm_upgrade_node_config.md
	modified:   content/en/docs/reference/setup-tools/kubeadm/generated/kubeadm_upgrade_plan.md
	modified:   content/en/docs/reference/setup-tools/kubefed/kubefed.md

Untracked files:
  (use "git add <file>..." to include in what will be committed)

	content/en/docs/reference/setup-tools/kubefed/kubefed_init.md
	content/en/docs/reference/setup-tools/kubefed/kubefed_join.md
	content/en/docs/reference/setup-tools/kubefed/kubefed_options.md
	content/en/docs/reference/setup-tools/kubefed/kubefed_unjoin.md
	content/en/docs/reference/setup-tools/kubefed/kubefed_version.md

no changes added to commit (use "git add" and/or "git commit -a")
```

My environment: Python 2.7 / Ubuntu 18.04 LTS

If anyone desires to do some additonal UAT against my test:
https://quirky-lamport-9bc79f.netlify.com/docs/reference/command-line-tools-reference
https://quirky-lamport-9bc79f.netlify.com/docs/reference/kubectl/kubectl
https://quirky-lamport-9bc79f.netlify.com/docs/reference/setup-tools/kubeadm/generated/
https://quirky-lamport-9bc79f.netlify.com/docs/reference/setup-tools/kubefed/kubefed

/LGTM

File: content/en/docs/contribute/generate-ref-docs/kubernetes-components.md [hunk 3]
```diff
@@ -118,134 +118,79 @@ need to work with someone who can set the label and milestone for you.
 
 The `update-imported-docs` tool performs these steps:
 
-1. Clone the `kubernetes/kubernetes` repository.
-1. Run several scripts under `kubernetes/kubernetes/hack`. These scripts
-   generate Markdown files and place the files under `kubernetes/kubernetes/docs`.
+1. Clone the related repositories specified in a configuration file. For the
```
[the remark this answers is not in the document]
Comment (tengqm) [reply content/en/docs/contribute/generate-ref-docs/kubernetes-components.md]: Look into the update-imported-docs directory, there are a few YAML files there to use.

File: content/en/docs/contribute/generate-ref-docs/kubernetes-components.md [hunk 4]
```diff
@@ -118,134 +118,79 @@ need to work with someone who can set the label and milestone for you.
...
-
-Open `<web-base>/update-imported-docs/config.yaml` for editing.
+## Customizing the config file
 
-Set the value of `branch` to the Kubernetes release that you want to document.
-For example, if you want to generate docs for the Kubernetes 1.9 release,
-set `branch` to `release-1.9`.
+Open `<web-base>/update-imported-docs/reference.yaml` for editing.
+Do not change the content for the `generate-command` entry unless you know what
+you are doing.
```
[the remark this answers is not in the document]
Comment (tengqm) [reply content/en/docs/contribute/generate-ref-docs/kubernetes-components.md]: The process for generating reference docs for kube components is a little bit tricky due to several facts:

- kubernetes/kubernetes is no longer checking in generated files which are needed for reference docs generation;
- `go get kubernetes/kubernetes` no longer works due to the same reason;
- reference docs generator is using some github Go packages which may conflict with kubernetes code base;
- there is no elegant solution today to manage go dependencies well.

The `generate-command` was carefully polished to make things work under these constraints. That is why we put a "heavy-handed" warning there.

The command is not supposed to be changed without being thoroughly tested in the reference-docs repo. Introducing a `include` directive in YAML is possible but I am not sure about the benefit we gain from doing that.

File: update-imported-docs/update-imported-docs [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+        except Exception as ex:
+            print("[Error] failed in writing target file '%s': %s"
+                  "" % (dst, str(ex)))
+            continue
+
+
+def main():
+    """The main entry of the program."""
+    if len(sys.argv) < 2:
+        print("[Error] Please specify a config file")
```
[the remark this answers is not in the document]
Comment (tengqm) [reply update-imported-docs/update-imported-docs]: Please check the `update-imported-docs` subdirectory.

File: content/en/docs/contribute/generate-ref-docs/kubernetes-components.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (kbhawkey) [reply content/en/docs/contribute/generate-ref-docs/kubernetes-components.md]: Hi, @tengqm. Thanks for the explanation. I'll dig a bit further into the scripts and configuration files.

File: content/en/docs/contribute/generate-ref-docs/kubernetes-components.md (hunk 2 above)
Comment (kbhawkey) [reply content/en/docs/contribute/generate-ref-docs/kubernetes-components.md]: Hi, @tengqm. Sorry, I am coming back to the generate-command. I am still trying to figure out the process.
- Specifying an external script for generate-command seems useful, especially if this script can be used in multiple places in reference.yaml.
- Will the federation component defined in reference.yaml use the generate-command?

File: content/en/docs/contribute/generate-ref-docs/kubernetes-components.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (kbhawkey) [reply content/en/docs/contribute/generate-ref-docs/kubernetes-components.md]: @tengqm, OK. The examples on this page refer to reference.yaml? I think it helps to be explicit.

**Comment by steveperry-53** [comment]
I'm going to say LGTM in the sense that this effort seems like it is in good hands. I think @chenopis should be the one to give a more official LGTM.

**Comment by jimangel** [comment]
Putting the "bat signal" up for @chenopis - any thoughts? :)

File: content/en/docs/contribute/generate-ref-docs/kubernetes-components.md (hunk 2 above)
Comment (tengqm) [reply content/en/docs/contribute/generate-ref-docs/kubernetes-components.md]: @kbhawkey Not sure how triggering works. Some examples would be helpful for feasibility evaluation.

**Comment by chenopis** [comment]
I'm going through the steps and testing it out now. Please standby...

File: content/en/docs/contribute/generate-ref-docs/kubernetes-components.md [hunk 6]
```diff
@@ -118,134 +118,79 @@ need to work with someone who can set the label and milestone for you.
...
-## Setting the branch
+## Customizing the config file
 
-Open `<web-base>/update-imported-docs/config.yaml` for editing.
-
-Set the value of `branch` to the Kubernetes release that you want to document.
-For example, if you want to generate docs for the Kubernetes 1.9 release,
-set `branch` to `release-1.9`.
+Open `<web-base>/update-imported-docs/reference.yaml` for editing.
+Do not change the content for the `generate-command` entry unless you know what
```
Comment (chenopis) [inline content/en/docs/contribute/generate-ref-docs/kubernetes-components.md]: "...unless you know what you are doing" is somewhat vague. Instead, how about "Do not change the content for the `generate-command` entry unless you understand what it is doing and need to change the specified release branch."

File: content/en/docs/contribute/generate-ref-docs/kubernetes-components.md [hunk 7]
```diff
@@ -118,134 +118,79 @@ need to work with someone who can set the label and milestone for you.
...
-you can run the `update-imported-docs` tool:
+After having reviewed and/or customized the `reference.yaml` file, you can run
+the `update-imported-docs` tool:
 
 ```shell
-cd <web-base>
-go get ./update-imported-docs
-go run update-imported-docs/update-imported-docs.go
+cd <web-base>/update-imported-docs
+./update-imported-docs reference.yml
```
Comment (chenopis) [inline content/en/docs/contribute/generate-ref-docs/kubernetes-components.md]: I'm getting the `ImportError: No module named yaml` error as well. And depending on what OS the user is on, there are different ways to install pyyaml. This is a potential friction point.

File: content/en/docs/contribute/generate-ref-docs/kubernetes-components.md [hunk 8]
```diff
@@ -118,134 +118,79 @@ need to work with someone who can set the label and milestone for you.
 
 The `update-imported-docs` tool performs these steps:
```
[the remark this answers is not in the document]
Comment (tengqm) [reply content/en/docs/contribute/generate-ref-docs/kubernetes-components.md]: Done

File: content/en/docs/contribute/generate-ref-docs/kubernetes-components.md (hunk 6 above)
Comment (tengqm) [reply content/en/docs/contribute/generate-ref-docs/kubernetes-components.md]: Done.

File: content/en/docs/contribute/generate-ref-docs/kubernetes-components.md (hunk 7 above)
Comment (tengqm) [reply content/en/docs/contribute/generate-ref-docs/kubernetes-components.md]: Okay, updating the script to print error message if PyYAML is not installed.

Comment (tengqm) [reply content/en/docs/contribute/generate-ref-docs/kubernetes-components.md]: Also updated the README.md by making this a requirement.

**Pull request merged**